### PR TITLE
Stop blacklisting Go 1.22+, drop Go < 1.21 support, use Go 1.22 in CI

### PR DESCRIPTION
**Pull request by kolyshkin**
High level overview:
 - remove nsenter compile restriction since Go 1.22.4 and Go 1.23 can now be used. THIS PART NEEDS TO BE BACKPORTED TO 1.1.
 - switch to Go 1.22 in CI whether possible (everywhere except Ubuntu 20.04 and CentOS 7 and 8).
 - set minimum requirement to Go 1.21
 - rm some old code.

See individual commits for details.

### CODE REVIEW

File: README.md
Comment (rata) [inline README.md:35]: maybe like this?
```suggestion
, Go 1.23 or Go >= 1.22.4.
```

**Review by rata — APPROVED** [review]
LGTM, thanks!

It took me a while to test locally that the go version dance was fine, but it indeed seem fine :)

File: README.md
Comment (kolyshkin) [reply README.md:35]: So, Go 1.22.4 would work, but the issue is, I can't disable the kludge in `libcontainer/nsenter/nsenter_go122.go` for Go >= 1.22.4, there's just no way to do it. I mention this problem in the first commit as well in this PR description.

I mean, we could add some kind of a hack to Makefile which checks go version and disables this...

Comment (kolyshkin) [reply README.md]: OK, ended up adding a hack to Makefile to check for go1.22 version, and removed the README addition (since if runc can't be compiled it prints a self-explanatory error).

PTAL

File: Makefile
Comment (AkihiroSuda) [inline Makefile:27]: `GGO` seems a typo

Comment (kolyshkin) [reply Makefile]: Thanks! fixed.

**Comment by kolyshkin** [comment]
We'll probably need to port the first 3 patches to release-1.1 (so that it will work with Go 1.22).

**Comment by kolyshkin** [comment]
> btw, it seems the ubuntu builds are stalled.

This PR drops go 1.20.x tests, but they are marked as required in repo config -- this is why they are shown as "Expected -- waiting ...". I will fix the repo config accordingly (rm go 1.20.x, require go 1.22.x) once this PR will be ready to be merged (I guess once Go 1.21.4 is out).

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers wdyt?

File: .github/workflows/test.yml
Comment (AkihiroSuda) [inline .github/workflows/test.yml:34]: Maybe add "TODO: enable this again when Go 1.22.4 is available"

File: README.md
Comment (AkihiroSuda) [inline README.md:30]: Maybe add some note about Go 1.22

**Review by AkihiroSuda — APPROVED** [review]

File: Makefile
Comment (rata) [inline Makefile:29]: Super nit, we are not using c++, so CFLAGS seem more appropriate (go passes both to C programs, though)

```suggestion
	export CGO_CFLAGS+="-D GO1_22_WORKS=1"
```

**Review by rata — APPROVED** [review]
LGTM

File: Makefile
Comment (kolyshkin) [reply Makefile:29]: I guess you're mixing up CPPFLAGS (for C PreProcessor) and CXXFLAGS (C++).

File: .github/workflows/test.yml
Comment (kolyshkin) [reply .github/workflows/test.yml:34]: done, thanks!

**Comment by kolyshkin** [comment]
@lifubang @cyphar PTAL

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
CentOS tests seem failing

**Comment by thaJeztah** [comment]
> CentOS tests seem failing

For CentOS stream 8, that may be expected; it's EOL since May 31 (just got hit by that in some CI pipeline that als had it in the Matrix)

**Comment by AkihiroSuda** [comment]
> > CentOS tests seem failing
> 
> For CentOS stream 8, that may be expected; it's EOL since May 31 (just got hit by that in some CI pipeline that als had it in the Matrix)

I guess we now have to replace CentOS with Alma, Rocky, or Oracle Linux?

**Review by mrunalp — APPROVED** [review]

**Comment by lifubang** [comment]
This one should be merged after go 1.22.4 has released.

File: Makefile
Comment (rata) [reply Makefile:29]: Oh, yes, sorry. `.cpp` being used as file extension for c++ projects really confuses me with the CPPFLAGS :-D

**Comment by kolyshkin** [comment]
Go 1.22.4 is out, this PR is rebased with the following change:

```diff
diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 4c40b47b..7155c163 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -30,11 +30,6 @@ jobs:
         criu: ["", "criu-dev"]
         dmz: ["", "runc_nodmz"]
         exclude:
-          # Go 1.22 + glibc < 2.32 + nsenter does not play well (issue 4233),
-          # thus nsenter package won't compile. Disable ubuntu 20.04 + go 1.22.
-          # TODO: remove this once Go 1.22.4 is out.
-          - os: ubuntu-20.04
-            go-version: 1.22.x
           # Disable most of criu-dev jobs, as they are expensive
           # (need to compile criu) and don't add much value/coverage.
           - criu: criu-dev
```

**Review by rata — APPROVED** [review]
LGTM.

The ubuntu 20.04 build is failing, probably 1.22.4 didn't make it on that action (the error is the C error about the version)

**Comment by kolyshkin** [comment]
Ah, `actions/setup-go` still installs go 1.22.3 (given go 1.22.x). They always lag behind (see https://github.com/actions/setup-go/issues/407).

Have to wait a few days I guess (or specify the version number explicitly, which I don't want to).

**Comment by kolyshkin** [comment]
@lifubang PTAL (except for pinning go in ci to 1.22.4 this is all good, and we can revert this patch later)

**Comment by AkihiroSuda** [comment]
Needs rebase

File: Makefile
Comment (AkihiroSuda) [inline Makefile:30]: I guess we can now just block Go 1.22.0-1.22.3 ?

Comment (kolyshkin) [reply Makefile:30]: You mean, block it from the top-level Makefile, and remove nsenter_go122 entirely? Makes sense

Comment (AkihiroSuda) [reply Makefile:30]: Yes

Comment (lifubang) [reply Makefile:30]: But for some libct users may need this file?

Comment (lifubang) [reply Makefile:30]: How about write some code to support all go 1.22 versions?

Comment (lifubang) [reply Makefile:30]: Please see #4310, feel free to close it if we don't want to support go 1.22.3, I think if we can support all go versions, it will be better.

Comment (AkihiroSuda) [reply Makefile:30]: Let's just skip 1.22.3, for simplicity

Comment (lifubang) [reply Makefile:30]: But for libct users, such as hashicorp/nomad, if they don’t write such method in Makefile, they will get an error even in go 1.22.4.

Comment (kolyshkin) [reply Makefile:30]: OTOH we are checking glibc version in nsenter_go122 and it's somewhat hard to do from the Makefile.

Comment (kolyshkin) [reply Makefile:37]: Leaving the current hack in place, added some extra comments.

Comment (kolyshkin) [reply Makefile:37]: > if we don't want to support go 1.22.3, I think if we can support all go versions, it will be better.

To me it doesn't make much sense to support those older versions. We can just require Go 1.22 >= 1.22.4, it's a decent requirement.

> But for libct users, such as hashicorp/nomad, if they don’t write such method in Makefile, they will get an error even in go 1.22.4.

That's right. 

Ideally, I'd like to do it all in one place -- that is, `nsenter_go122.go`. Unfortunately, I can't find a way to determine go's patch version from there.

Comment (kolyshkin) [reply Makefile:37]: > I can't find a way to determine go's patch version from there.

... during compile time.

Comment (rata) [reply Makefile:37]: Is it so bad to not handle the case at all, then? I mean, not do anything for go 1.22, just document in the function/package that if using go 1.22, you need at least go 1.22.4 (or glibc and explain it all) and be done with it.

Go 1.22.4 also includes security fixes, so hopefully people upgrade faster.

A runc release will at least take 1-2 more weeks I think, so go 1.22.4 should be out for a few weeks already by then.

Comment (rata) [reply Makefile:37]: IMHO, I'd leave this PR without handling the 1.22 case, we can discuss what to do for 1.22 #4310

Comment (lifubang) [reply Makefile:37]: I agree, because go 1.22.4 has released, we think that people should swith to go 1.22.4, they will not use go 1.22.3, especially when running CI.

Comment (kolyshkin) [reply Makefile:37]: Yeah, maybe we should just stop blacklisting go and instead add a note that for Go 1.22 one should use Go 1.22.4+.

**Review by rata — APPROVED** [review]
LGTM

File: README.md
Comment (kolyshkin) [reply README.md:32]: thanks! I need more coffee this morning :)

**Comment by kolyshkin** [comment]
Assuming that by the time 1.2.0 is out, everyone will use Go 1.22.x  >= 1.22.4, I just dropped the restriction (and added a note to README).

I think this is ready to be merged. WDYT @cyphar?

File: README.md
Comment (lifubang) [inline README.md]: In line 30, I have tested it and found that runc can’t compile in go 1.19 anymore, we should change it to ‘go 1.20 or higher’.

**Review by lifubang — APPROVED** [review]

File: README.md
Comment (lifubang) [reply README.md]: Note: This is an error review if we’re doing review commit by commit.

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**
